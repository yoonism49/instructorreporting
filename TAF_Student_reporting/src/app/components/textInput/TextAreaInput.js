// @flow weak

import React, {
  PureComponent
}                     from 'react';
import PropTypes      from 'prop-types';


class TextAreaInput extends PureComponent {
  static propTypes = {
    label:    PropTypes.string.isRequired,
    id:       PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    value:    PropTypes.string.isRequired,
    nbrows:   PropTypes.number,
    onChange: PropTypes.func.isRequired,
    delay:    PropTypes.number
  };

  static defaultProps = {
    nbrows: 3,
    delay: 200
  };

  state = { stateValue: '' };

  timer = null;

  componentWillReceiveProps(nextProps) {
    const { stateValue } = this.state;
    const { value } = nextProps;

    if ((value !== stateValue) && stateValue.length === 0) {
      this.setState({stateValue: value});
    }
  }

  componentWillUnmount() {
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
  }

  render() {
    const {label, id, nbrows} = this.props;
    const { stateValue } = this.state;

    return (
      <div className="form-group">
        <label
          htmlFor={id}
          className="control-label">
          {label}
          </label>
        <div>
          <textarea
            className="form-control"
            rows={nbrows}
            id={id}
            // value={stateValue}
            defaultValue={stateValue}
            // onChange={this.handlesOnChange} // IE11 misses some keys entered... yes I know what you think...
            onInput={this.handlesOnChange}
          />
        </div>
      </div>
    );
  }

  handlesOnChange = (event: any) => {
    event.preventDefault();
    this.setState({stateValue: event.target.value});
    this.setTimerBeforeCallback(event.target.value);
  }
  // hack to prevent bad user xp when huge forms and callback each onChange to parent or store like redux:
  setTimerBeforeCallback(value) {
    const { onChange, delay } = this.props;

    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }

    this.timer = setTimeout(
      () => onChange(value),
      delay
    );
  }
}

export default TextAreaInput;
