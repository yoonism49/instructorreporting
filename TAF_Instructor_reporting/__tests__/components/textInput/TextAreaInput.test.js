
'use strict';

import React              from 'react';
import TextAreaInput      from '../../../src/app/components/textInput/TextAreaInput';
import renderer           from 'react-test-renderer'; // needed both for snpashot testing but also to prevent errors from enzyme

describe('TextAreaInput component', () => {
  const mockProps = {
    label:    'test',
    id:       1,
    value:    '',
    onChange: () => (true)
  };

  it('renders as expected', () => {
    const component = renderer.create(
      <div>
        <TextAreaInput {...mockProps}/>
      </div>
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
});
