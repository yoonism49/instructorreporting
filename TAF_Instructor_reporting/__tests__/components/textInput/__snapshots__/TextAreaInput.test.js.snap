// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaInput component renders as expected 1`] = `
<div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor={1}
    >
      test
    </label>
    <div>
      <textarea
        className="form-control"
        defaultValue=""
        id={1}
        onInput={[Function]}
        rows={3}
      />
    </div>
  </div>
</div>
`;
